[JIRA KEYCLOAK-831] SAML 2.0 HTTP-Artifact binding - AuthNRequest

AlistairDoswald opened the pull request:
This is a partial implementation of Artifact binding for SAML, handling only the login (AuthNRequest) 

- Keycloak will recognize an HTTP-Artifact message in the AuthNRequest
- Can force artifact binding and a return address in the GUI
- Added tests in integration-arquillian (including builer for the client)

The rest (artifact binding for LogoutRequest and answering an artifact with an ArtifactResolve) will be implemented within the next two weeks (normally). I understand that the PR may not be accepted as it is not the full implementation, but:
1. The feature works, even if it is not yet the full implementation
2. I wouldn't mind a review change to be sure I'm going in the right direction.

Commit:
Added Artifact Binding for Authnrequest

- Keycloak will recognize an HTTP-Artifact message in the Authnrequest
- Can force artifact binding and a return address in the GUI
- Added tests in integration-arquillian (including builer for the client)

Commit:
Added handling of "artifact" messages

- Added http-artifact binding
- Added handling of artifact messages
- Added in code and GUI entry for remote artifact resolution servier
- Added some corrections and modifications for artifact authnrequest
- Added to and modified test framework to handle artifact and artifact resolve test cases
- Added tests to cover handling of artifact messages

Commit:
Added a couple of assertions in test case

AlistairDoswald commented:
I have a few of comments about this pull request of mine:
1. I pushed two commits to my fork, but I didn't realise that they would automatically become part or the pull request (sorry).
2. However, I have completed the artifact binding on a branch (full feature). How should I handle this? Should I push everything to this pull request or remove it and make a new one which correctly describes the scope? I already asked this question on the mailing list, but have no answer
3.  Since the feature is complete, I've also written modifications to the documentation. I will do a pull request on the documentation once I know what I should do for point 2.

hmlnarik commented:
@AlistairDoswald Thank you for the contribution. Sadly, current priorities and bandwidth does not allow us to work on this PR at this moment. Therefore this PR is currently marked on hold. We will be happy to continue the work after 4.x is wrapped up.
Thank you for understanding.

AlistairDoswald commented:
@hmlnarik I understand. However, could you give me an indication of what I should do for the partial feature/full feature issue? Should I merge the complete artifact-binding code into this PR, or should I make a new single commit PR with the complete feature and close this one (and then the new one can also be put on hold)

Commit:
Full artifact binding protocol

- Added code for full implmentation of artifact binding (also included automatic tests)
- Updated code to current state of keycloak
- Correct * imports and assert* to assertThat as asked by nmlnarik

AlistairDoswald commented:
@hmlnarik it took me a while since I had other priorities and there seemed to be no rush, but I've now added the full artifact-binding for SAML. 

Concerning your remarks I've removed the `*` imports and changed all asserts to `assertThat`, but not the suggestion for the `ClientAttributeUpdater` as I'm not sure to what part of the code you are referring to. 

For the "isPostBinding" and "forcePostBinding", currently I'm using the get/post state to determine if the returned messages with the artifact should be transmitted with a redirect or a POSTed for, so it's not exactly `redirect OR post OR artifact` but actually `redirect OR post OR artifact+redirect OR artifact+post`, so I think that in the code it makes sense to keep the code as-is. However, in the admin console it may make sense to have, instead two of the two current switches, a "force protocol response" switch which leaves the choice between redirect/post/artifact. In this case it may not be necessary to let the user decide if he wants the artifact to transmitted with a redirect or a POSTed form.

I've also written some documentation, so I'll create a PR on the keycloak-documentation to go with this PR.

Please let me know your thoughts once you have the time to look at this PR.

hmlnarik commented:
@AlistairDoswald Thanks a lot! I'll have a close look soon (in few weeks) I hope. Just briefly:

- ClientAttributeUpdater is related to the tests. Check e.g. https://github.com/keycloak/keycloak/blob/3c44e6c3773d708796b8abee6891d7ae98ac0cf8/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/IdpInitiatedLoginTest.java#L79 for usage - you can replace code in https://github.com/keycloak/keycloak/blob/00ac3d16f677e3e9d9c4825351b591b4cd9797e0/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/ArtifactBindingTest.java#L172-L178 with that.
- It seems some unrelated changes got in the PR, could you please check the last commit?
- Re the configuration, I believe the best would be start a thread on keycloak-dev after we unfreeze the feature development.

AlistairDoswald commented:
@hmlnarik Thanks for the information, and OK for waiting to start a thread on keycloak-dev. 

For the commit, I see the problem. I had synchronised my fork to work with the latest version of the keycloak code and mishandled the commit and push, resulting in all updated files being part of the commit. Unfortunatly, I'm not sure how to clean up this mess. I can easily cleanup the state of my local repository, so that I only have the diff between my code and the upstream/master, but I'm not sure what will happen to this PR if I force push to origin. I'll try to figure out if there's a safer way to clean up the PR.

Commit:
rolled back keycloak to before artifact binding

Commit:
merge 4.8.3.Final into master

Commit:
Added artifact-binding into 4.8.3.Final

AlistairDoswald commented:
@hmlnarik I tried out a few options to clean up the mess I had made, but in the end this was the cleanest I found which preserved the history. I first reverted the code to the state of 4.8.3.Final, and then I patched the Artifact-Binding code into it. The result is that the latest commit (Added artifact-binding into 4.8.3.Final 84a91ba) has only the files that are modified between Keycloak 4.8.3.Final and Keycloak 4.8.3.Final with artifact-binding.

AlistairDoswald commented:
@hmlnarik I've just noticed that the hold tag on this PR was actually removed 2 weeks ago (and that keycloak 5 has been released). You mentioned in a previous comment that I should start a thread about the configuration once feature dev has been unfrozen... would now be a good time? 

Also, I see that there's now some conflicts with the master, should I address this, or keep the code clean?

Finally, we've been deploying this code in a pre-prod environment to test it against Service Providers from several different vendors that support/require artifact binding, and an issue was found during the import of a client from its metadata with an optional parameter. Should I commit the correction into this PR? Or is it better addressed at a later time?

hmlnarik commented on services/src/main/java/org/keycloak/protocol/saml/SamlService.java:
```diff
@@ -732,14 +888,168 @@ public AuthenticationSessionModel getOrCreateLoginSessionForIdpInitiatedSso(Keyc
+                }
+            }
+        }
+        //check if it is the logout message
```
The change of the deferred model of user session removal is unfortunately not justifiable in this case. However there is way to build the SAML art in a way that it contains all the information needed to reconstruct logout response. SAML art has to be prepared so that it is not forgerable. This can be done by actually creating a JWT token (yes, the same from OIDC) upon creating the Artifact response with limited time validity. This would be validated there via `TokenVerifier` and based on its values, the LogoutResponse would be sent (or an error in case it would be invalid).

If you prefer, this part can be written by myself, feel free to let me know.

hmlnarik commented on services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:
```diff
@@ -103,6 +103,7 @@
     public static final String SET_REDIRECT_URI_AFTER_REQUIRED_ACTIONS= "SET_REDIRECT_URI_AFTER_REQUIRED_ACTIONS";
     public static final String END_AFTER_REQUIRED_ACTIONS = "END_AFTER_REQUIRED_ACTIONS";
     public static final String INVALIDATE_ACTION_TOKEN = "INVALIDATE_ACTION_TOKEN";
+    public static final String DEFER_CLEANUP = "DEFER_CLEANUP";
```
The change of the deferred model of user session removal is unfortunately not justifiable in this case.

hmlnarik reviewed (changes requested):
@AlistairDoswald Thank you for your contribution and patience. It is a nice piece of the code, regardless of my comments which do not dispute code quality but rather are based on broader context and limitations.

Thank you very much for getting the changes actually to a single commit only. I appreciate your work to get the artifact binding work to be the last commit. That helped a lot with the review. To move this forward, I would like to ask you to:

1. Split the changes to the support of ARTIFACT logic (https://issues.jboss.org/browse/KEYCLOAK-9992) and the UI (https://issues.jboss.org/browse/KEYCLOAK-9995) into separate commits and pull requests. It would be helpful to first have the KEYCLOAK-9992 part implemented before the UI.

2. There will be need to update the format of the SAMLArt to eliminate the need to walk through the user/client sessions. This is possible, as the `urn:oasis:names:tc:SAML:2.0:artifact-04` is not mandatory per specification

3. Deferring the deletion of the user session to after `finishLogout` is sadly not justifiable in this case. This can be replaced by preparing the SAML art in the way that it is only recognized as valid token by the running instance, e.g. by using a JWT token. This part is going to be tricky and I will be happy to implement if you prefer it once the other comments would be addressed.

Thank you!

AlistairDoswald commented:
@hmlnarik thanks for the in depth review. I've taken the time to check your comments and make some verifications. I'm still interested in moving this code and PR forward, but I unfortunately won't be able to start the code modifications before the 25th of April.

Now, to address your comments:

1. This isn't a problem for me. If this is OK for you, for https://issues.jboss.org/browse/KEYCLOAK-9992,  I'd also like to take the opportunity to move the code to the latest tagged version of keycloak, and also integrate a couple of bug fixes that we discovered from our pre-prod testing.
2. You're correct in saying that **urn:oasis:names:tc:SAML:2.0:artifact-04** is not mandatory per specification, however two of the SPs I've tested the code against expect it, and will raise an error if the artifact has any other format. I'm concerned that changing the format it'll make the code less compatible with SPs that support artifact binding.  By now the artifact-binding code has been tested against SPs from 10 different vendors, and I'll try to contact them to find out whatever libraries or implementation they are using have any requirement on artifact format. In the comments on the code you mentioned that the iterative solution will not scale to a clustered environment. I don't have any experience with the problems that may arise with clustered environments, so could you elaborate a bit more on what you meant? Basically, will the system not work at all in a clustered environment, or will there be some harsh performance penalties? Because in the former case the **urn:oasis:names:tc:SAML:2.0:artifact-04** may be unable to be supported at all, but in the latter it could be possible to support both the **urn:oasis:names:tc:SAML:2.0:artifact-04** and the other artifact format you describe.
3. I was never very happy with the deferred logout, but I couldn't see any other way to handle the situation. I'll see if I can work out your jwt suggestion, but I'll be happy to have some help otherwise.

I don't see any problems with your other comments on code snippets, so I'll be making those changes

hmlnarik commented:
@AlistairDoswald Thank you for your willingness to continue!

1. That's great, I agree. Let's replace this PR with one for KEYCLOAK-9992. Feel free to close this and open a new PR or replace the contents of this PR with the new code, whatever suits you better.
2. The way out would be to introduce a SPI that would resolve (`getArtifact` method), and create and store (most of the `buildArtifactAuthenticatedResponse`) necessary information about the artifacts; possibly some other methods I do not see at this moment too. The first implementation would use Keycloak's internal format, there it would be easy to avoid deferred logout. (I do appreciate deferred logout was a good idea to overcome the current limitations btw.) There could be an implementation conforming to `urn:oasis:names:tc:SAML:2.0:artifact-04` format later, yet let's start with a more straightforward one.
_Re the session handling:_ user / client sessions are distributed across the cluster, so any iteration across all sessions involves network overhead that is proportional to the number of the sessions. Hence such iterations should be prevented. When strictly needed, one has to use a predicate like `org.keycloak.models.sessions.infinispan.stream.UserSessionPredicate`. However I believe this implementation should prevent this happen. The situation gets even more sensitive performance-wise when using cross data center setup.
To understand the context, it would be useful to know the vendors of the failing implementations. If that is possible, could you please share those?
3. Absolutely, feel free to start an implementation, I'll be happy to help!

stianst commented:
Closing this as out of date

hmlnarik commented:
I'm handling this PR to get it to mergeable state.

hmlnarik commented:
Closing in favor of #7756
